File: cloud-driver-multiplatform/cloud-driver-multiplatform-java/src/main/java/de/lino/cloud/platform/rest/api/dto/Dtos.java
package de.lino.cloud.platform.rest.api.dto;

import java.util.Map;

/**
 * Plain request/response shapes mirrored 1:1 against the server's REST contract - see
 * {@code CloudRestExtension}/{@code DefaultRestFactory} in cloud-driver for the authoritative
 * field names. Deliberately not shared code with the server module: the desktop client only
 * ever talks HTTP, so it has no dependency on cloud-driver-api at all.
 */
public final class Dtos {

    /** Not instantiable - a pure namespace for the nested record types below. */
    private Dtos() {
    }

    /**
     * Body for {@code POST /auth/login} and {@code POST /auth/register} - both read the same
     * {@code {"username", "password"}} shape server-side. The field is named {@code username}
     * to match the server's {@code AuthService} contract exactly, even though the value passed
     * in is always an email address.
     */
    public record AuthRequest(String username, String password) {
    }

    /**
     * Body for {@code POST /auth/register/confirm} - step two of registration, submitting the
     * 6-digit code {@code POST /auth/register} e-mailed. {@code username} is the same email
     * address passed to {@link AuthRequest}, named the same way for the same reason.
     */
    public record ConfirmRegistrationRequest(String username, String code) {
    }

    /**
     * Response of {@code POST /auth/login}, {@code POST /auth/register/confirm}, {@code
     * POST /auth/reset-password/confirm}, and {@code POST /auth/refresh} - every one of the
     * server's token-issuing routes returns this same shape. {@code refreshToken} is a
     * longer-lived, opaque, single-use token (rotated on every {@code POST /auth/refresh} call) a
     * client can exchange for a fresh pair once {@code token} expires, without asking the user to
     * log in again - see {@link ApiClient#refresh}.
     */
    public record AuthResponse(String token, String refreshToken) {
    }

    /**
     * Body for {@code POST /auth/refresh} and {@code POST /auth/logout} - both take a bare refresh
     * token, previously returned in an {@link AuthResponse}.
     */
    public record RefreshRequest(String refreshToken) {
    }

    /** Body for {@code POST /auth/reset-password} - starts a password reset for the account under {@code username}. */
    public record RequestPasswordResetRequest(String username) {
    }

    /**
     * Body for {@code POST /auth/reset-password/confirm} - submits the 6-digit code {@code
     * POST /auth/reset-password} e-mailed, plus the caller's chosen {@code newPassword}.
     */
    public record ConfirmPasswordResetRequest(String username, String code, String newPassword) {
    }

    /**
     * Body for {@code POST /auth/change-email} - bearer-gated (unlike every other request record
     * in this class): the account being changed is the caller's own, resolved server-side from
     * its bearer token, not from anything in this body. Starts an e-mail change for the
     * already-authenticated caller; {@code newEmail} is not live yet - only {@link
     * ApiClient#confirmEmailChange} actually applies it.
     */
    public record ChangeEmailRequest(String newEmail) {
    }

    /**
     * Body for {@code POST /auth/change-email/confirm} - submits the 6-digit code {@code
     * POST /auth/change-email} e-mailed to the pending new address.
     */
    public record ConfirmChangeEmailRequest(String code) {
    }

    /**
     * Response of {@code POST /auth/register} ({@code 202 Accepted}) - registration is not
     * complete yet at this point, this only acknowledges that a verification code was e-mailed.
     */
    public record MessageResponse(String message) {
    }

    /**
     * Shape of the object {@code POST /files} and {@code GET /files/{id}} return on success -
     * <b>not</b> {@code GET /files}'s own listing, which returns {@link StoredFileSummaryResponse}
     * instead (see that record's Javadoc). Mirrors {@code StoredFile}'s Gson-serialized fields;
     * {@code checksum} is left out entirely here since the client only needs {@code
     * fileId}/{@code fileName}/{@code contentBase64} to download and save a file. {@code
     * folderId} is not a {@code StoredFile} field at all - the server merges it in from the
     * file's {@code StoredFileOwnership} row (see {@code DefaultRestFactory#toJsonObject}) -
     * {@code null} for a file at the root.
     */
    public record StoredFileResponse(
            String fileId,
            String fileName,
            String contentType,
            String contentBase64,
            boolean contentCompressed,
            long createdAtEpochMilli,
            long updatedAtEpochMilli,
            String folderId
    ) {
    }

    /**
     * Shape of one entry in the {@code GET /files} response array - a file's descriptive fields
     * plus its folder placement, deliberately without content (mirrors the server's {@code
     * StoredFileSummary}). Fetch a specific file's full content afterwards via {@code
     * GET /files/{id}} ({@link ApiClient#downloadFile}), which returns a {@link
     * StoredFileResponse} instead.
     *
     * <p>{@code scanStatus} mirrors the server's content-scan verdict for this file - one of
     * {@code "CLEAN"}/{@code "PENDING"}/{@code "FLAGGED"}, treated here as an opaque string. A
     * file uploaded on a deployment with no content-scanning extension running never carries
     * anything but {@code "CLEAN"}.
     */
    public record StoredFileSummaryResponse(
            String fileId,
            String fileName,
            String contentType,
            long sizeBytes,
            long createdAtEpochMilli,
            long updatedAtEpochMilli,
            String folderId,
            String scanStatus
    ) {
    }

    /**
     * Shape of one entry in the {@code GET /folders} response array, and of the object
     * {@code POST /folders}/{@code PUT /folders/{id}} return on success. Mirrors {@code
     * Folder}'s Gson-serialized fields; {@code ownerId} is included even though the client
     * never needs to act on it, purely because it's part of the server's actual JSON shape.
     * {@code color} (added for per-folder display color) mirrors {@code Folder#getColor()} -
     * an opaque, client-defined string (e.g. {@code "BLUE"}), {@code null} if never explicitly
     * set (a client falls back to its own default).
     */
    public record FolderResponse(
            String folderId,
            String ownerId,
            String name,
            String parentFolderId,
            long createdAtEpochMillis,
            long modifiedAtEpochMillis,
            String color
    ) {
    }

    /** Body for {@code POST /folders} - {@code parentFolderId} {@code null} creates a top-level folder. */
    public record CreateFolderRequest(String name, String parentFolderId) {
    }

    /**
     * Body for {@code PUT /folders/{id}} - a full replace of both fields (matching {@code PUT}'s
     * whole-resource-replace semantics), not a partial patch; {@code parentFolderId} {@code null}
     * moves the folder to the top level.
     */
    public record UpdateFolderRequest(String name, String parentFolderId) {
    }

    /** Body for {@code PUT /folders/{id}/color} - {@code color} {@code null} clears it back to "unset". */
    public record UpdateFolderColorRequest(String color) {
    }

    /** Body for {@code PUT /files/{id}/folder} - {@code folderId} {@code null} moves the file back to the root. */
    public record MoveFileRequest(String folderId) {
    }

    /** Body for {@code PUT /files/{id}/rename} - {@code fileName} is the file's new display name. */
    public record RenameFileRequest(String fileName) {
    }

    /**
     * Body for {@code PUT /cloudUsers/theme} (added 2026-09-04) - syncs the caller's light/dark
     * theme choice to their account so it follows them across devices, replacing what used to be
     * a per-device local setting. {@code themeMode} is treated as an opaque string server-side
     * (e.g. {@code "LIGHT"}/{@code "DARK"} - the actual enum lives client-side); {@code null}
     * clears the stored preference back to "unset".
     */
    public record UpdateThemeRequest(String themeMode) {
    }

    /**
     * Body for {@code POST /files/upload-url} - the first step of a presigned, direct-to-client
     * upload. {@code sizeBytes} is checked against quota now and again (against the real uploaded
     * size) at {@link CompleteUploadRequest}.
     */
    public record BeginUploadUrlRequest(String fileName, long sizeBytes, String folderId) {
    }

    /**
     * Response from {@code POST /files/upload-url} - {@code requiredHeaders} must be replayed
     * exactly on the client's own {@code PUT} to {@code uploadUrl}, or the object store rejects
     * the request's signature.
     */
    public record BeginUploadUrlResponse(String fileId, String uploadUrl, Map<String, String> requiredHeaders, long expiresAtEpochMillis) {
    }

    /**
     * Body for {@code POST /files/{id}/complete-upload} - the second step of a presigned upload.
     * No {@code sizeBytes} field here: the server always re-reads the real size from the object
     * store itself, never trusting the client's declared size a second time.
     */
    public record CompleteUploadRequest(String fileName, String checksumSha256, String folderId) {
    }

    /** Response from {@code GET /files/{id}/download-url} - the client {@code GET}s {@code downloadUrl} directly, bypassing this server. */
    public record BeginDownloadUrlResponse(String downloadUrl, long expiresAtEpochMillis) {
    }

    /** Body Javalin's default error responses use ({@code BadRequestResponse} etc. all share this shape). */
    public record ErrorResponse(String title) {
    }

    /**
     * Mirrors the server's {@code CursorPage<T>} response envelope - what {@code GET /files}/
     * {@code GET /folders} return instead of a bare array once a caller passes {@code ?limit=}
     * (see {@code DefaultRestFactory#toPageEnvelope}). {@code nextCursor} is {@code null} once
     * there is no next page; pass it back as the next call's {@code ?cursor=} otherwise.
     *
     * @param items      up to the requested page size
     * @param nextCursor the cursor to request the next page with, or {@code null} if this was the last page
     * @param <T>        the element type of one page - {@link StoredFileSummaryResponse} or {@link FolderResponse}
     */
    public record Page<T>(java.util.List<T> items, String nextCursor) {
    }

    /**
     * Shape of the object {@code GET /cloudUsers/{id}} returns - mirrors {@code CloudUser}'s
     * Gson-serialized fields. {@code timeStamp} is set once, when the {@code CloudUser} record is
     * first created (account confirmation time), so it doubles as the account's creation
     * timestamp - see {@code CloudUser}'s own Javadoc server-side. {@code maxBytesToUpload}/{@code
     * currentUploadedBytes} back the Dashboard's s3storage-quota display - see {@code
     * ICloudUser#getMaxBytesToUpload()}/{@code #getCurrentUploadedBytes()} server-side for what
     * each actually tracks. {@code themeMode} (added 2026-09-04) is this account's synced
     * light/dark theme preference - {@code null} if never explicitly set, in which case a client
     * falls back to its own default; update it via {@code PUT /cloudUsers/theme} (see {@link
     * de.lino.cloud.platform.rest.api.ApiClient#updateThemePreference}).
     */
    public record CloudUserResponse(String authUserId, long timeStamp, long maxBytesToUpload, long currentUploadedBytes, String themeMode) {
    }

    /**
     * Response of {@code GET /auth/me} (bearer-gated) - the caller's own account id, email
     * address, and admin flag. Used by the desktop client to decide whether to show its Admin
     * sidebar entry at all, since there is otherwise no way for a client to learn this without
     * probing an admin-gated route and interpreting a {@code 403}.
     */
    public record MeResponse(String authUserId, String emailAddress, boolean isAdmin) {
    }

    /**
     * Shape of one entry in the {@code GET /admin/authUsers} response array (admin-gated) -
     * mirrors {@code AuthUser}'s Gson-serialized fields, keeping only what the desktop client's
     * read-only Admin panel actually displays; extra fields present in the server's actual JSON
     * (e.g. {@code passwordHash}) are simply ignored by Gson on deserialization into this smaller
     * shape.
     */
    public record AuthUserResponse(String id, String emailAddress, boolean isAdmin) {
    }

    /**
     * Shape of one entry in the {@code GET /admin/audit-log} response array (admin-gated) -
     * mirrors the server's {@code AuditLogEntryResponse}: {@code actorEmail} is already resolved
     * server-side from the underlying actor id, {@code null} for an action with no identified
     * actor (e.g. a failed login against an unknown address).
     */
    public record AuditLogEntryResponse(long timestampEpochMillis, String action, String actorEmail, String targetId) {
    }

    /**
     * Response of {@code GET /files/shared-by-me/count} - how many of the caller's own files
     * currently have at least one active share (the owner-side count, distinct from {@code
     * GET /files/shared-with-me}'s grantee-side listing). Backs the desktop app's Dashboard
     * "Shared files" stat card.
     */
    public record SharedByMeCountResponse(int count) {
    }

    /**
     * Response of {@code GET /admin/metrics} (admin-gated) - mirrors the server's {@code
     * MetricsSnapshot} field-for-field, read in-process off {@code cloud-driver-extensions-metrics}'s
     * Prometheus registry rather than a separate scrape call. {@code 404}/{@code
     * ApiException(503, ...)} if that extension isn't running on this deployment at all - the
     * desktop client should treat that as "metrics unavailable", not a bug.
     */
    public record MetricsSnapshotResponse(long uploadsSucceeded, long uploadsFailed, long uploadsQueued,
                                           long uploadQuotaRejections, long pendingUploadQueueDepth,
                                           Map<String, Long> extensionsByStatus) {
    }

    /**
     * Body for {@code POST /files/{id}/share} and {@code POST /folders/{id}/share} - grants
     * {@code granteeEmail}'s account access. {@code permissionLevel} ({@code "VIEW"}/{@code
     * "EDIT"}, {@code null} defaults to {@code "VIEW"} server-side) and {@code
     * expiresAtEpochMillis} ({@code null} means the grant never expires) are both optional -
     * omitting them (the original two-field shape this record used to be) behaves exactly as
     * before. {@code EDIT} only ever takes effect on a direct file share, never a folder share or
     * a folder-inherited grant.
     */
    public record ShareRequest(String granteeEmail, String permissionLevel, Long expiresAtEpochMillis) {
    }

    /**
     * Response of {@code GET /cloudUsers/exists?email=<address>} - whether <em>any</em> account is
     * registered under that address. Backs the desktop app's Share dialog live-checking a typed
     * grantee address before submitting a share.
     */
    public record EmailExistsResponse(boolean exists) {
    }

    /**
     * Shape of one entry in the {@code GET /files/shared-with-me} response array (added 2026-09-02,
     * replacing a bare {@link StoredFileSummaryResponse} array) - mirrors the server's {@code
     * SharedFileSummary}: the file's own descriptive fields plus {@code ownerEmail}, the email
     * address of the account that shared it.
     */
    public record SharedFileSummaryResponse(StoredFileSummaryResponse file, String ownerEmail) {
    }

    /**
     * Shape of one entry in the {@code GET /folders/shared-with-me} response array (added
     * 2026-09-02, replacing a bare {@link FolderResponse} array) - mirrors the server's {@code
     * SharedFolderSummary}, the same "pair with the sharing account's email" shape as {@link
     * SharedFileSummaryResponse}.
     */
    public record SharedFolderSummaryResponse(FolderResponse folder, String ownerEmail) {
    }

    /**
     * Shape of one entry in the {@code GET /files/trash} response array (added 2026-09-02,
     * replacing a bare {@link StoredFileSummaryResponse} array) - mirrors the server's {@code
     * TrashedFileSummary}: the file's own descriptive fields plus {@code purgeAtEpochMillis}, when
     * it becomes eligible for permanent removal under the server's configured trash retention
     * window.
     */
    public record TrashedFileSummaryResponse(StoredFileSummaryResponse file, long purgeAtEpochMillis) {
    }

    /**
     * Shape of one entry in the {@code GET /folders/trash} response array (added 2026-09-02,
     * replacing a bare {@link FolderResponse} array) - mirrors the server's {@code
     * TrashedFolderSummary}, the same "pair with a purge-eligibility timestamp" shape as {@link
     * TrashedFileSummaryResponse}.
     */
    public record TrashedFolderSummaryResponse(FolderResponse folder, long purgeAtEpochMillis) {
    }

    /**
     * Response of {@code GET /folders/{id}/shared-contents} (added 2026-09-02) - the non-trashed
     * files/subfolders directly inside a folder reached via ownership or a share, mirroring the
     * server's {@code SharedFolderContents}. Backs the desktop app's "browse into a shared folder"
     * and "download this shared folder" actions.
     */
    public record SharedFolderContentsResponse(java.util.List<StoredFileSummaryResponse> files, java.util.List<FolderResponse> subfolders) {
    }

    /**
     * Shape of one entry in the {@code GET /search} response array - a matched file's id/name plus
     * its folder placement. Content is never included, matching {@link StoredFileSummaryResponse}'s
     * own "listing never carries content" convention.
     */
    public record SearchResultResponse(String storedFileId, String fileName, String folderId) {
    }

    /**
     * Shape of one entry in the {@code GET /files/{id}/public-link} response array, and of the
     * object {@code POST /files/{id}/public-link} returns on success - an unauthenticated,
     * read-only link to one file's content. {@code expiresAtEpochMillis} is {@code null} if the
     * link never expires.
     */
    public record PublicFileLinkSummaryResponse(String token, long createdAtEpochMillis, Long expiresAtEpochMillis) {
    }

    /** Body for {@code POST /files/{id}/public-link} - {@code expiresAtEpochMillis} {@code null} creates a link that never expires. */
    public record CreatePublicFileLinkRequest(Long expiresAtEpochMillis) {
    }

    /**
     * Shape of one entry in the {@code GET /files/{id}/versions} response array - a captured prior
     * version of a file's content, identified by its 1-based {@code versionNumber}. Deliberately
     * carries no {@code fileName}/{@code contentType} - the server's own version-history record has
     * neither, since a version only ever preserves content, not the file's descriptive metadata at
     * the time it was captured.
     */
    public record FileVersionSummaryResponse(int versionNumber, long capturedAtEpochMillis, long sizeBytes) {
    }

    /**
     * Shape of one entry in the {@code GET /activity}/{@code GET /files/{id}/activity}/
     * {@code GET /folders/{id}/activity} response envelopes' {@code items} array - a single
     * audit-log entry. The server's own audit event carries more fields ({@code id}, {@code
     * actorAuthUserId}, {@code metadata}) that this record deliberately omits, since Gson simply
     * ignores any JSON field with no matching record component on deserialization.
     */
    public record ActivityEntryResponse(long timestampEpochMillis, String action, String targetId) {
    }

    /** Body for {@code POST /webhooks} - {@code eventTypes} names which event types (e.g. {@code "FILE_UPLOADED"}) this subscription receives. */
    public record RegisterWebhookRequest(String url, java.util.List<String> eventTypes) {
    }

    /**
     * Shape of one entry in the {@code GET /webhooks} response array - a registered webhook
     * subscription, without its secret (only ever handed back once, at creation - see {@link
     * WebhookSubscriptionCreatedResponse}).
     */
    public record WebhookSubscriptionSummaryResponse(String id, String url, java.util.Set<String> eventTypes, long createdAtEpochMillis) {
    }

    /**
     * Shape of the object {@code POST /webhooks} returns on success - the same fields as {@link
     * WebhookSubscriptionSummaryResponse} plus {@code secret}, the value used to sign every
     * delivery's {@code X-Webhook-Signature} header. Shown only this once; store it immediately,
     * since no later call ever returns it again.
     */
    public record WebhookSubscriptionCreatedResponse(String id, String url, java.util.Set<String> eventTypes, String secret, long createdAtEpochMillis) {
    }

    /**
     * Shape of one entry in the {@code GET /webhooks/deliveries} response array - one attempt to
     * deliver one event to one subscription, successful or not.
     */
    public record WebhookDeliveryAttemptResponse(String webhookId, String eventType, String targetId, long attemptedAtEpochMillis,
                                                  int attemptNumber, boolean succeeded, Integer responseStatusCode) {
    }

}

File: cloud-driver-multiplatform/cloud-driver-multiplatform-swift/Sources/CloudDriverSwift/Dtos.swift
import Foundation

// Plain request/response shapes mirrored 1:1 against the server's REST contract - see
// cloud-driver's `DefaultRestFactory`/`CloudRestExtension` for the authoritative field names,
// and `cloud-driver-maven`'s `Dtos.java` for the JVM-side twin of this file. Deliberately
// hand-kept-in-sync rather than shared code: this app only ever talks HTTP, the same boundary
// `cloud-driver-maven` itself enforces against the server modules.
//
// Every type/member below is `public` - this file lives in the `CloudDriverSwift` package (moved
// out of cloud-driver-platforms-mobile, 2026-09-07), so anything the app (or another consumer)
// needs to read has to cross the module boundary explicitly. Request (`Encodable`) types are only
// ever constructed inside `APIClient.swift`, which lives in this same package/module - so their
// memberwise initializers don't strictly need to be `public` for this package's own use, but are
// marked `public` anyway (via an explicit `init`, since Swift's synthesized memberwise init is
// never `public` even on a `public` type) for any future external consumer of this library.

/// The `{"items", "nextCursor"}` envelope `GET /files`/`GET /folders` return when called with
/// `?limit=` (cursor pagination, mirrors the server's `CursorPage<T>` and
/// `cloud-driver-maven`'s `Dtos.Page<T>`/`cloud-driver-platforms-desktop`'s
/// `Dtos.Page<T>`). `nextCursor` is `nil` once the last page has been reached. Swift's `Codable`
/// handles a generic type directly - no `TypeToken`/reflection dance needed the way the JVM
/// clients require.
public struct Page<T: Decodable>: Decodable {
    public let items: [T]
    public let nextCursor: String?
}

public struct AuthRequest: Encodable {
    public let username: String
    public let password: String

    public init(username: String, password: String) {
        self.username = username
        self.password = password
    }
}

public struct ConfirmRegistrationRequest: Encodable {
    public let username: String
    public let code: String

    public init(username: String, code: String) {
        self.username = username
        self.code = code
    }
}

public struct RequestPasswordResetRequest: Encodable {
    public let username: String

    public init(username: String) {
        self.username = username
    }
}

public struct ConfirmPasswordResetRequest: Encodable {
    public let username: String
    public let code: String
    public let newPassword: String

    public init(username: String, code: String, newPassword: String) {
        self.username = username
        self.code = code
        self.newPassword = newPassword
    }
}

public struct RefreshRequest: Encodable {
    public let refreshToken: String

    public init(refreshToken: String) {
        self.refreshToken = refreshToken
    }
}

public struct CreateFolderRequest: Encodable {
    public let name: String
    public let parentFolderId: String?

    public init(name: String, parentFolderId: String?) {
        self.name = name
        self.parentFolderId = parentFolderId
    }
}

public struct MoveFileRequest: Encodable {
    public let folderId: String?

    public init(folderId: String?) {
        self.folderId = folderId
    }
}

/// Body for `PUT /files/{id}/rename` - a separate route from `MoveFileRequest`'s, matching the
/// server's own separate handling (renaming rewrites the file's own entity, unlike a move).
public struct RenameFileRequest: Encodable {
    public let fileName: String

    public init(fileName: String) {
        self.fileName = fileName
    }
}

/// Body for `POST /files/upload-url` - the first step of a presigned, direct-to-client upload.
/// `sizeBytes` is checked against quota now and again (against the real uploaded size) at
/// `CompleteUploadRequest`.
public struct BeginUploadUrlRequest: Encodable {
    public let fileName: String
    public let sizeBytes: Int64
    public let folderId: String?

    public init(fileName: String, sizeBytes: Int64, folderId: String?) {
        self.fileName = fileName
        self.sizeBytes = sizeBytes
        self.folderId = folderId
    }
}

/// Response from `POST /files/upload-url` - `requiredHeaders` must be replayed exactly on this
/// app's own `PUT` to `uploadUrl`, or the object store rejects the request's signature.
public struct BeginUploadUrlResponse: Decodable {
    public let fileId: String
    public let uploadUrl: String
    public let requiredHeaders: [String: String]
    public let expiresAtEpochMillis: Int64
}

/// Body for `POST /files/{id}/complete-upload` - the second step of a presigned upload. No
/// `sizeBytes` field here: the server always re-reads the real size from the object store itself.
public struct CompleteUploadRequest: Encodable {
    public let fileName: String
    public let checksumSha256: String
    public let folderId: String?

    public init(fileName: String, checksumSha256: String, folderId: String?) {
        self.fileName = fileName
        self.checksumSha256 = checksumSha256
        self.folderId = folderId
    }
}

/// Response from `GET /files/{id}/download-url` - this app `GET`s `downloadUrl` directly, bypassing the server.
public struct BeginDownloadUrlResponse: Decodable {
    public let downloadUrl: String
    public let expiresAtEpochMillis: Int64
}

/// Body for `PUT /folders/{id}` - a full replace of both fields (matching `PUT`'s
/// whole-resource-replace semantics), used here to change only `parentFolderId` while carrying
/// the folder's existing `name` through unchanged.
public struct UpdateFolderRequest: Encodable {
    public let name: String
    public let parentFolderId: String?

    public init(name: String, parentFolderId: String?) {
        self.name = name
        self.parentFolderId = parentFolderId
    }
}

/// Body for `POST /auth/change-email` - bearer-gated: the account being changed is the caller's
/// own, resolved server-side from its token, never from this body.
public struct ChangeEmailRequest: Encodable {
    public let newEmail: String

    public init(newEmail: String) {
        self.newEmail = newEmail
    }
}

/// Body for `POST /auth/change-email/confirm`.
public struct ConfirmChangeEmailRequest: Encodable {
    public let code: String

    public init(code: String) {
        self.code = code
    }
}

/// Body for `POST /files/{id}/share` and `POST /folders/{id}/share` - grants `granteeEmail`'s
/// account access. `permissionLevel` (`"VIEW"`/`"EDIT"`, `nil` defaults to `VIEW` server-side) and
/// `expiresAtEpochMillis` (`nil` never expires) are both optional, so an existing call passing only
/// `granteeEmail` keeps compiling and behaving exactly as before either field existed.
public struct ShareRequest: Encodable {
    public let granteeEmail: String
    public let permissionLevel: String?
    public let expiresAtEpochMillis: Int64?

    public init(granteeEmail: String, permissionLevel: String? = nil, expiresAtEpochMillis: Int64? = nil) {
        self.granteeEmail = granteeEmail
        self.permissionLevel = permissionLevel
        self.expiresAtEpochMillis = expiresAtEpochMillis
    }
}

public struct MessageResponse: Decodable {
    public let message: String
}

/// Response shape returned by every token-issuing route (login, register/confirm,
/// reset-password/confirm, refresh).
public struct AuthResponse: Decodable {
    public let token: String
    public let refreshToken: String
}

/// Shape of one entry in `GET /files`'s response array, and of `POST /files`'s own response body
/// (upload responses carry no content - see cloud-driver's "Large-file upload/download streaming"
/// notes). Deliberately without content; fetch bytes separately via `GET /files/{id}/content`.
///
/// `scanStatus` mirrors the server's content-scan verdict (`"CLEAN"`/`"PENDING"`/`"FLAGGED"`) -
/// treated as opaque text here, not a Swift enum, the same "kept as a plain string, real meaning
/// lives server-side" convention this file already applies to `FolderResponse.color`.
public struct StoredFileSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { fileId }
    public let fileId: String
    public let fileName: String
    public let contentType: String
    public let sizeBytes: Int64
    public let createdAtEpochMilli: Int64
    public let updatedAtEpochMilli: Int64
    public let folderId: String?
    public let scanStatus: String
}

/// Shape of one entry in `GET /folders`'s response array, and of what `POST /folders`/`PUT
/// /folders/{id}` return on success. `color` mirrors the server's `Folder#getColor()` - an
/// opaque, client-defined string (e.g. `"BLUE"`), `nil` if never explicitly set (this app falls
/// back to its own default - see `FolderColorOption.forName(_:)` in Theme.swift).
public struct FolderResponse: Decodable, Identifiable, Hashable {
    public var id: String { folderId }
    public let folderId: String
    public let ownerId: String
    public let name: String
    public let parentFolderId: String?
    public let createdAtEpochMillis: Int64
    public let modifiedAtEpochMillis: Int64
    public let color: String?
}

/// Body for `PUT /folders/{id}/color` - a separate route from `UpdateFolderRequest`'s, matching
/// the server's own separate handling (recoloring never touches name/parent).
public struct UpdateFolderColorRequest: Encodable {
    public let color: String?

    public init(color: String?) {
        self.color = color
    }
}

/// Shape of one entry in `GET /files/shared-with-me`'s response array - a file another account
/// shared with the caller, paired with that account's email (mirrors the server's
/// `SharedFileSummary`).
public struct SharedFileSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { file.fileId }
    public let file: StoredFileSummaryResponse
    public let ownerEmail: String
}

/// Shape of one entry in `GET /folders/shared-with-me`'s response array - the folder counterpart
/// to `SharedFileSummaryResponse` (mirrors the server's `SharedFolderSummary`).
public struct SharedFolderSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { folder.folderId }
    public let folder: FolderResponse
    public let ownerEmail: String
}

/// Response of `GET /folders/{id}/shared-contents` - the non-trashed files/subfolders directly
/// inside a folder reached via ownership or a share (mirrors the server's `SharedFolderContents`).
/// A share on an ancestor folder covers browsing into any descendant, so this same call also
/// works for a subfolder reached by navigating deeper into an already-shared folder.
public struct SharedFolderContentsResponse: Decodable {
    public let files: [StoredFileSummaryResponse]
    public let subfolders: [FolderResponse]
}

/// Shape of one entry in `GET /files/trash`'s response array - a soft-deleted file (via `DELETE
/// /files/{id}`) paired with when it becomes eligible for permanent removal under the server's
/// configured trash retention window (mirrors the server's `TrashedFileSummary`).
public struct TrashedFileSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { file.fileId }
    public let file: StoredFileSummaryResponse
    public let purgeAtEpochMillis: Int64
}

/// Shape of one entry in `GET /folders/trash`'s response array - the folder counterpart to
/// `TrashedFileSummaryResponse` (mirrors the server's `TrashedFolderSummary`).
public struct TrashedFolderSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { folder.folderId }
    public let folder: FolderResponse
    public let purgeAtEpochMillis: Int64
}

/// Response of `GET /auth/me` (bearer-gated) - the caller's own account id/email/admin flag.
public struct MeResponse: Decodable {
    public let authUserId: String
    public let emailAddress: String
    public let isAdmin: Bool
}

/// Response of `GET /cloudUsers/{id}` - mirrors `CloudUser`'s Gson-serialized fields server-side.
public struct CloudUserResponse: Decodable {
    public let authUserId: String
    public let timeStamp: Int64
    public let maxBytesToUpload: Int64
    public let currentUploadedBytes: Int64
}

/// Body Javalin's default error responses use (`BadRequestResponse` etc. all share this shape).
public struct ErrorResponse: Decodable {
    public let title: String
}

/// Response of `GET /cloudUsers/exists?email=` - answers "does any account exist under this
/// address", not scoped to the caller's own account. Used to live-check a grantee's address
/// before sharing, the same way `cloud-driver-platforms-desktop`'s `ShareDialog` does.
public struct EmailExistsResponse: Decodable {
    public let exists: Bool
}

/// Shape of one entry in `GET /search`'s response array - a filename/content match against the
/// caller's own account, scoped per-account (a search never matches another account's files).
public struct SearchResultResponse: Decodable, Identifiable, Hashable {
    public var id: String { storedFileId }
    public let storedFileId: String
    public let fileName: String
    public let folderId: String?
}

/// Shape of one entry in `GET /search/semantic`'s response array.
///
/// `SearchResultResponse` plus a `score` - the cosine similarity between the query and the file,
/// higher being closer. Kept as its own type rather than folding an optional score onto the
/// keyword shape, because the two answer different questions: a keyword search either matched or
/// it did not, whereas every semantic result matched to *some* degree and the score is what makes
/// the list interpretable at all.
public struct SemanticSearchResultResponse: Decodable, Identifiable, Hashable {
    public var id: String { storedFileId }
    public let storedFileId: String
    public let fileName: String
    public let folderId: String?
    public let score: Double
}

/// One member of a `DuplicateFileGroupResponse`.
public struct DuplicateFileEntryResponse: Decodable, Identifiable, Hashable {
    public var id: String { storedFileId }
    public let storedFileId: String
    public let fileName: String
    public let folderId: String?
}

/// Shape of one entry in `GET /files/duplicates`'s response array - files the server considers
/// near-identical in meaning, plus the group's weakest pairwise similarity.
///
/// **Not the same as byte-identical.** The server deduplicates identical content exactly and
/// invisibly; this is a similarity judgement meant for a human to review, so a client must present
/// it as a suggestion and must never delete anything on its own.
public struct DuplicateFileGroupResponse: Decodable, Hashable {
    public let files: [DuplicateFileEntryResponse]
    public let similarity: Double
}

/// Shape of one entry in `GET /files/{id}/tags`'s response array.
///
/// `confidence` is a *relative* similarity against a fixed label vocabulary, not a calibrated
/// probability - do not render it to a user as a percentage of correctness.
public struct TagSuggestionResponse: Decodable, Identifiable, Hashable {
    public var id: String { tag }
    public let tag: String
    public let confidence: Double
}

/// Body for `POST /files/{id}/public-link` - `expiresAtEpochMillis` `nil` creates a link that
/// never expires.
public struct CreatePublicFileLinkRequest: Encodable {
    public let expiresAtEpochMillis: Int64?

    public init(expiresAtEpochMillis: Int64? = nil) {
        self.expiresAtEpochMillis = expiresAtEpochMillis
    }
}

/// Shape of one entry in `GET /files/{id}/public-link`'s response array, and of what `POST
/// /files/{id}/public-link` returns on success - an unauthenticated, read-only link to one file's
/// content, reachable via `GET /public/files/{token}` with no bearer token at all.
public struct PublicFileLinkSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: String { token }
    public let token: String
    public let createdAtEpochMillis: Int64
    public let expiresAtEpochMillis: Int64?
}

/// Shape of one entry in `GET /files/{id}/versions`'s response array - a previously-captured
/// version of a file's content (captured automatically whenever `PUT /files/{id}/content`
/// overwrites it). Deliberately narrower than `StoredFileSummaryResponse` - a version carries no
/// `fileName`/`contentType` of its own in this response, only what changed release to release.
public struct FileVersionSummaryResponse: Decodable, Identifiable, Hashable {
    public var id: Int { versionNumber }
    public let versionNumber: Int
    public let capturedAtEpochMillis: Int64
    public let sizeBytes: Int64
}

/// One entry in an activity feed page (`GET /activity`, `GET /files/{id}/activity`, `GET
/// /folders/{id}/activity`) - a narrowed view of the server's audit-log entry, decoding only the
/// fields this app actually displays (`Decodable` silently ignores the rest, e.g. `actorAuthUserId`/
/// `metadata`). `id` is synthesized from timestamp+target rather than decoded, since the server's
/// own entry id isn't one of the fields read here.
public struct ActivityEntryResponse: Decodable, Identifiable, Hashable {
    public var id: String { "\(timestampEpochMillis)-\(targetId ?? "")" }
    public let timestampEpochMillis: Int64
    public let action: String
    public let targetId: String?
}
